controllers/cluster_controller_test.go

```diff
@@ -211,4 +219,170 @@ var _ = Describe("Cluster Reconciler", func() {
 			return instance.Finalizers
 		}, timeout).Should(BeEmpty())
 	})
+
+	It("Should successfully set Status.ControlPlaneInitialized on the cluster object if controlplane is ready", func() {
+		cluster := &clusterv1.Cluster{
+			ObjectMeta: metav1.ObjectMeta{
+				GenerateName: "test-",
+				Namespace:    "default",
+			},
+		}
+
+		Expect(k8sClient.Create(ctx, cluster)).To(BeNil())
+		key := client.ObjectKey{Name: cluster.Name, Namespace: cluster.Namespace}
+		defer k8sClient.Delete(ctx, cluster)
+
+		// Wait for reconciliation to happen.
+		Eventually(func() bool {
+			if err := k8sClient.Get(ctx, key, cluster); err != nil {
+				return false
+			}
+			return len(cluster.Finalizers) > 0
+		}, timeout).Should(BeTrue())
+
+		machine := &clusterv1.Machine{
+			ObjectMeta: metav1.ObjectMeta{
+				GenerateName: "test-",
+				Namespace:    v1.NamespaceDefault,
+				Labels: map[string]string{
+					clusterv1.MachineControlPlaneLabelName: cluster.Name,
+					clusterv1.MachineClusterLabelName:      cluster.Name,
+				},
+			},
+			Spec: clusterv1.MachineSpec{
+				ProviderID: pointer.StringPtr("aws:///id-node-1"),
+				Bootstrap: clusterv1.Bootstrap{
+					Data: pointer.StringPtr(""),
+				},
+			},
+		}
+
+		Expect(k8sClient.Create(ctx, machine)).To(BeNil())
+		key = client.ObjectKey{Name: machine.Name, Namespace: machine.Namespace}
+		defer k8sClient.Delete(ctx, machine)
+
+		// wait for machine to be ready
+		Eventually(func() bool {
+			if err := k8sClient.Get(ctx, key, machine); err != nil {
+				return false
+			}
+			return len(machine.Finalizers) > 0
+		}, timeout).Should(BeTrue())
+
+		// patch machine noderef
+		patchHelper, err := patch.NewHelper(machine, k8sClient)
+		Expect(err).NotTo(HaveOccurred())
+
+		machine.Status.NodeRef = &v1.ObjectReference{}
+		Expect(patchHelper.Patch(ctx, machine)).Should(BeNil())
+
+		// Assertion
+		key = client.ObjectKey{Name: cluster.Name, Namespace: cluster.Namespace}
+		Eventually(func() bool {
+			if err := k8sClient.Get(ctx, key, cluster); err != nil {
+				return false
+			}
+			return cluster.Status.ControlPlaneInitialized
+		}, timeout).Should(BeTrue())
+	})
 })
+
+func TestClusterReconciler_MachineToCluster(t *testing.T) {
+	cluster := &clusterv1.Cluster{
+		TypeMeta: metav1.TypeMeta{
+			Kind: "Cluster",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "test-cluster",
+			Namespace: "test",
+		},
+		Spec:   clusterv1.ClusterSpec{},
+		Status: clusterv1.ClusterStatus{},
+	}
+
+	m1 := &clusterv1.Machine{
```

Comment:
Could you please name m1, m2, m3 more descriptively, such as controlPlaneMachineWithNodeRef, controlPlaneMachineWithoutNodeRef, nonControlPlaneMachine?

Reply:
resolved! Actually i took reference from here :stuck_out_tongue:  https://github.com/kubernetes-sigs/cluster-api/blob/11f4fb13223a40afe848782eaa2f175012a27002/controllers/machineset_controller_test.go#L206-L249